src/class/dfu/dfu_mode_device.h

```diff
@@ -0,0 +1,131 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * This file is part of the TinyUSB stack.
+ */
+
+#ifndef _TUSB_DFU_MODE_DEVICE_H_
+#define _TUSB_DFU_MODE_DEVICE_H_
+
+#include "common/tusb_common.h"
+#include "device/usbd.h"
+#include "dfu.h"
+
+#ifdef __cplusplus
+  extern "C" {
+#endif
+
+
+//--------------------------------------------------------------------+
+// Application Callback API (weak is optional)
+//--------------------------------------------------------------------+
+// Invoked when a reset is received to check if firmware is valid
+bool tud_dfu_mode_firmware_valid_check_cb(void);
+
+// Invoked when the device must reboot to dfu runtime mode
+void tud_dfu_mode_reboot_to_rt_cb(void);
+
+// Invoked during initialization of the dfu driver to set attributes
+// Return byte set with bitmasks:
+//   DFU_FUNC_ATTR_CAN_DOWNLOAD_BITMASK
+//   DFU_FUNC_ATTR_CAN_UPLOAD_BITMASK
+//   DFU_FUNC_ATTR_MANIFESTATION_TOLERANT_BITMASK
+//   DFU_FUNC_ATTR_WILL_DETACH_BITMASK
+// Note: This should match the USB descriptor
+uint8_t tud_dfu_mode_init_attrs_cb(void);
+
+// Invoked during a DFU_GETSTATUS request to get for the string index
+// to the status description string table.
+TU_ATTR_WEAK uint8_t tud_dfu_mode_get_status_desc_table_index_cb(void);
+
+// Invoked during a USB reset
+// Lets the app perform custom behavior on a USB reset.
+// If not defined, the default behavior remains the DFU specification of
+// Checking the firmware valid and changing to the error state or rebooting to runtime
+// Note: If used, the application must perform the reset logic for all states.
+// Changing the state to APP_IDLE will result in tud_dfu_mode_reboot_to_rt_cb being called
+TU_ATTR_WEAK void tud_dfu_mode_usb_reset_cb(uint8_t rhport, dfu_mode_state_t *state);
+
+// Invoked during a DFU_GETSTATUS request to set the timeout in ms to use
+// before the subsequent DFU_GETSTATUS requests.
+// The purpose of this value is to allow the device to tell the host
+// how long to wait between the DFU_DNLOAD and DFU_GETSTATUS checks
+// to allow the device to have time to erase, write memory, etc.
+// ms_timeout is a pointer to array of length 3.
+// Refer to the USB DFU class specification for more details
+TU_ATTR_WEAK void tud_dfu_mode_get_poll_timeout_cb(uint8_t *ms_timeout);
+
+// Invoked when a DFU_DNLOAD request is received
+// This should start a timer for ms_timeout ms
+// When the timer has elapsed, tud_dfu_runtime_poll_timeout_done must be called
+// NOTE: This callback should return immediately, and not implement the delay
+// internally, as this will hold up the class stack from receiving any packets
+// during the DFU_DNLOAD_SYNC and DFU_DNBUSY states
+void tud_dfu_mode_start_poll_timeout_cb(uint8_t *ms_timeout);
+
+// Must be called when the poll_timeout has elapsed
+void tud_dfu_mode_poll_timeout_done(void);
```

Comment:
instead of using timer based, this API should be renamed to `tud_dfu_download_complete()` which is called when the flashing is complete by application (started by tud_dfu_mode_req_dnload_data_cb() ). From what I understand the poll timeout is just a suggestion for host to time its next DOWNLOAD request. There is no need to keep the timer on. If host decide to request before we actually complete, we can simply STALL the request to reject it.

Reply:
Refactored.
Please review b8e5885